src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -8004,21 +8004,21 @@ bool tryAsMemberOfSingleType(NamedTypeSymbol singleType, [NotNullWhen(true)] out
 
             (BoundExpression operand, Conversion conversion) = RemoveConversion(node, includeExplicitConversions: true);
             SnapshotWalkerThroughConversionGroup(node, operand);
-            TypeWithState operandType = VisitRvalueWithState(operand);
```

Comment:
I think the `Unsplit` should be kept for the general case. We should only have a split state for an expression of type `bool` (I will open a PR shortly to encode this invariant, which is surprisingly not enforced yet).
The general machinery of a conditional state is composed of a StateWhenTrue and StateWhenFalse, which are not suitable to represent a split state for any type but `bool`.
For the few scenarios where we need to keep track of a split state for a type other than `bool`, we use specialized machinery instead (typically introducing a variable `PossiblyConditionalState stateWhenNotNull`).

So I think we should restrict this change to only to error scenarios of conversions to bool (the operand has an error type and the target type is `bool`).

Reply:
> (the operand has an error type and the target type is `bool`).

It feels like it is also acceptable to keep the conditional state when both the source and target of the conversion are bool. In other words, let's keep the conditional state when the target type of the conversion is `bool`, and the source type is either `bool` or an error type.

Reply:
Done, but I had to also include target of bool? as it was failing a test (and I don't think it's wrong because it's still from bool or error)

Reply:
Could you please show the test that was failing when we unsplit for `bool?` and the difference in behavior? I think we'll want to understand that scenario very clearly before we could agree that a `bool?` source type should be unsplit.

Reply:
It was `NullCoalescing_ConditionalLeft`.

Specfically:

```cs
(bool?)(b && c1.M(x = 0)) ?? false
            ? x.ToString() // 1
            : x.ToString(); // 2
```
This now behaves the same as before this pr (both are warned)

And same for this one:

```cs
(bool?)((y = 0) is 0 && c1.M(x = 0)) ?? false
            ? x.ToString() + y.ToString() // 3
            : x.ToString() + y.ToString(); // 4
```
The actual expression before the cast is of type bool so it shouldn't be unsplit.

The problem I realised is this pr shouldn't change this test because whether or not this is true or false shouldn't change that we don't know if x or y are null.

But in order to maintain this, the target type also needs to allow for bool? target conversion to not unsplit.

Does that answer your question?

Reply:
oh sorry I got it backwards: when adding the restrictions, the test reverted until I allowed the bool?

I realised it was a mistake so let me remove that which should revert the test.

Reply:
done.